packages/cli/src/modules/insights/insights.service.ts

```diff
@@ -89,68 +135,139 @@ export class InsightsService {
+		}
+
+		await this.sharedWorkflowRepository.manager.transaction(async (trx) => {
+			const sharedWorkflows = await trx.find(SharedWorkflow, {
+				where: { workflowId: In([...workflowIdNames.keys()]), role: 'workflow:owner' },
+				relations: { project: true },
 			});
 
-			if (!metadata) {
-				// This can't happen, we just wrote the metadata in the same
-				// transaction.
-				throw new UnexpectedError(
-					`Could not find metadata for the workflow with the id '${ctx.workflowData.id}'`,
-				);
+			// Upsert metadata for the workflows that are not already in the cache or have
+			// different project or workflow names
+			const metadataToUpsert = sharedWorkflows.reduce((acc, workflow) => {
+				const cachedMetadata = this.cachedMetadata.get(workflow.workflowId);
+				if (
+					!cachedMetadata ||
+					cachedMetadata.projectId !== workflow.projectId ||
+					cachedMetadata.projectName !== workflow.project.name ||
+					cachedMetadata.workflowName !== workflowIdNames.get(workflow.workflowId)
+				) {
+					const metadata = new InsightsMetadata();
+					metadata.projectId = workflow.projectId;
+					metadata.projectName = workflow.project.name;
+					metadata.workflowId = workflow.workflowId;
+					metadata.workflowName = workflowIdNames.get(workflow.workflowId)!;
+
+					acc.push(metadata);
+				}
+				return acc;
+			}, [] as InsightsMetadata[]);
+
+			await trx.upsert(InsightsMetadata, metadataToUpsert, ['workflowId']);
+
+			const upsertMetadata = await trx.findBy(InsightsMetadata, {
+				workflowId: In(metadataToUpsert.map((m) => m.workflowId)),
+			});
+			for (const metadata of upsertMetadata) {
+				this.cachedMetadata.set(metadata.workflowId, metadata);
 			}
 
-			// success or failure event
-			{
-				const event = new InsightsRaw();
-				event.metaId = metadata.metaId;
-				event.type = status;
-				event.value = 1;
-				await trx.insert(InsightsRaw, event);
+			const events: InsightsRaw[] = [];
+			for (const event of insightsRawToInsertBuffer) {
+				const insight = new InsightsRaw();
+				const metadata = this.cachedMetadata.get(event.workflowId);
+				if (!metadata) {
+					// could not find shared workflow for this insight (not supposed to happen)
+					throw new UnexpectedError(
+						`Could not find shared workflow for insight with workflowId ${event.workflowId}`,
+					);
+				}
+				insight.metaId = metadata.metaId;
+				insight.type = event.type;
+				insight.value = event.value;
+
+				events.push(insight);
 			}
 
-			// run time event
-			if (fullRunData.stoppedAt) {
-				const value = fullRunData.stoppedAt.getTime() - fullRunData.startedAt.getTime();
-				const event = new InsightsRaw();
-				event.metaId = metadata.metaId;
-				event.type = 'runtime_ms';
-				event.value = value;
-				await trx.insert(InsightsRaw, event);
-			}
+			await trx.insert(InsightsRaw, events);
+		});
+	}
+
+	async flushEvents() {
+		// Prevent flushing if there are no events to flush and we are not forcing it
```

Comment:
What does the "and we are not forcing it" mean?

Reply:
My bad, old comment when I previously used a variable to force flush. Removing it...